[mono] add internal WebWorkerEventLoop utility class

lambdageek opened the pull request:
This is Part 3 of https://github.com/dotnet/runtime/issues/84489 - landing support for async JS interop on threadpool threads in multi-threaded WebAssembly.

Provides two pieces of functionality:

1. A keepalive token that can be used to prevent the current POSIX thread from terminating when it returns from its thread start function, or from an invocation from the JS event loop.  When the last keepalive token is destroyed (assuming Emscripten isn't keeping the thread alive for other reasons) it will terminate as if by calling `pthread_exit` and the webworker will be made available to other threads

2. A `HasUnsettledInteropPromises` property that peeks `_js_owned_object_table` to see if there are any promises created by the interop subsystem that have not been fulfilled or rejected yet.

lambdageek committed:
[mono] add internal WebWorkerEventLoop utility class

Provides two pieces of functionality:

1. A keepalive token that can be used to prevent the current POSIX
thread from terminating when it returns from its thread start
function, or from an invocation from the JS event loop.  When the last
keepalive token is destroyed (assuming Emscripten isn't keeping the
thread alive for other reasons) it will terminate as if by calling
`pthread_exit` and the webworker will be made available to other
threads

2. A `HasUnsettledInteropPromises` property that peeks
`_js_owned_object_table` to see if there are any promises created by
the interop subsystem that have not been fulfilled or rejected yet.

pavelsavara commented on src/mono/wasm/runtime/gc-handles.ts:
```diff
@@ -131,3 +131,9 @@ export function _lookup_js_owned_object(gc_handle: GCHandle): any {
+/// Called from the C# threadpool worker loop to find out if there are any
+/// unsettled JS promises that need to keep the worker alive
+export function mono_wasm_eventloop_has_unsettled_interop_promises(): boolean {
+    return _js_owned_object_table.size > 0;
```
This table also has all other types of objects in it, not just promises. We may need to track promises separately. The `JSHandle` still needs to be thread-unique, not just unique per type.

pavelsavara replied:
> > This table also has all other types of objects in it, not just promises.
> 
> I see. So it is probably better to just augment `_marshal_task_to_cs` to increment/decrement an unsettled promises counter.

yes, when in MT build.

> > The `JSHandle` still needs to be thread-unique, not just unique per type.
> 
> Sorry, I don't understand what this means.

The other implementation option could be to have table per type. But allocation of the JSHandle would have to be common.

lambdageek committed:
Use a per-thread unsettled promise count

for mono_wasm_eventloop_has_unsettled_interop_promises we
can't use the _js_owned_object_table size because it contains
other interop objects, not just promises

lambdageek commented on src/mono/wasm/runtime/marshal-to-cs.ts:
```diff
@@ -305,10 +308,17 @@ function _marshal_task_to_cs(arg: JSMarshalerArgument, value: Promise<any>, _?:
     setup_managed_proxy(holder, gc_handle);
 
+    if (monoWasmThreads)
+        addUnsettledPromise();
```
I'm not sure what we should do when we return a Task to a JS worker.

I'm having trouble imagining what a sample would look like that does this that doesn't also somehow return to C#.  (ie: I first `JSImport` something and that something is async and it does an `await` on a `JSExport` - so everything is alive as long as the c# is awaiting the JSImport-ed thing)

lambdageek replied:
Yea, in fact I tried something like that with setTimeout. 

I'm not sure if it's better to add something when we pass a js handle for a c# object, or just to add a public c# API to ensure the thread stays alive. It seems hard to guess what is the right thing.

pavelsavara replied:
The other type of "event" which should keep the thread alive is events from WebSocket. 
I guess JS side of the WS code could call internal `addUnsettledPromise()`.

Perhaps we need better name as it's not only about promises.

lambdageek replied:
> And so you would have to keep the thread alive until the last C# function proxy was GC'd by `FinalizationRegistry`.

I think this is probably what we need to do, then.  Both for C# functions and C# Tasks.  Maybe even all C# objects that get passed to JS.

If something in JS is holding on to a C# object of any kind, the C# threadpool worker has to stay alive because JS may trigger some C# Task to become unblocked and require executing on that webworker.

lambdageek replied:
@pavelsavara So did we decide something here?  Should I block the worker thread from exiting while it has any proxies that haven't been GC'd on the JS side?

pavelsavara replied:
To really free all proxies, worst case scenarios could involve multiple runs of both GCs. And I don't like it.

So, probably not on this PR. Let's create issue for it and link it from this place in code where we only count promises.
So that we have chance to experiment with different solutions separately and unblock the rest of the progress on MT.

lambdageek replied:
Created #85052

lambdageek committed:
remove old emscripten keepalive workaround hack

lambdageek committed:
Add a more general getter for JS interop to keep a webworker alive

And link to #85052 for more details

lambdageek committed:
fixup docs

lambdageek commented:
@pavelsavara anything else I should address in this one?

pavelsavara commented:
I'm ok about the code, but I don't understand the big picture.
I guess this PR doesn't have usage of `WebWorkerEventLoop` in it and it would be in the follow-up PR part 4 ?
Why is the `KeepaliveToken` instance or multiple instances. Would simple counter suffice ?

lambdageek commented:
> I'm ok about the code, but I don't understand the big picture. I guess this PR doesn't have usage of 
`WebWorkerEventLoop` in it and it would be in the follow-up PR part 4 ?

Yea it's used in the Part 4.

> Why is the `KeepaliveToken` instance or multiple instances. Would simple counter suffice ?

I wanted to make something that is harder to misuse than the emscripten `push`/`pop` API. Also to mirror the JS `setTimeout`/`clearTimeout` API - so every caller will get a unique keepalive token.  I kind of suspect this will eventually become a public API

pavelsavara commented:
Thanks for explanation. 
I see how this is useful, assuming there would be more than one push/pop per worker and that from external code.

> I kind of suspect this will eventually become a public API

I'm not convinced this should be public API. Once we have use case for it we could re-consider it. For now, LGTM

lambdageek commented:
debugger test timeout is https://github.com/dotnet/runtime/issues/84101
